### DISCUSSION TITLE
Adding Unique decorator on inheritting entity causes parent to acquire unique constraint

### DISCUSSION THREAD

**Discussion by Jcalcaldev** [mikro-orm/mikro-orm#5041] [General]
### Describe the bug

I currently have a parent `Identity` entity and 2 children: `GoogleIdentity` and `WalletIdentity`.

The parent class has a user manyToOne relation:

```
@Entity({
  discriminatorColumn: 'type',
  abstract: true,
  discriminatorMap: {
    google: 'GoogleIdentity',
    wallet: 'WalletIdentity',
  },
})
export class Identity extends BaseEntity {
  @PrimaryKey()
  public id: string = v4();

  @ManyToOne(() => User, { inversedBy: 'identities', onDelete: 'cascade' })
  public user: User;
  ...
```

If I attempt to add a @Unique decorator to one of the children like so:

```
@Entity({
  discriminatorValue: 'google',
})
@Unique({ properties: ['user'] })
export class GoogleIdentity extends Identity {
  @Property()
  public email: string;
  ...
```

And create my migrations, I'm getting this on the generated file:
`this.addSql('alter table "identity" add constraint "identity_type_user_id_unique" unique ("type", "user_id");');`

The issue here is that it's making all identity's user property unique, and my intended outcome is that Users can have many identities, but only one googleIdentity, and more than one walletIdentity.

User: 
```
@Entity()
export class User extends BaseEntity {
  @PrimaryKey()
  public id: string = v4();

  @OneToMany(() => Identity, 'user', { orphanRemoval: true })
  public identities: Collection<Identity> = new Collection<Identity>(this);
  ...
```

Is there any way to achieve the implementation I'm looking for at the entity level, without having to resort to the service level in order to achieve the unique constraints I'm looking for?

### Reproduction

* 

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

5.8.1

### Node.js version

18.18.0

### Operating system

macOS Sonoma 14.2

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
I guess you could have some conditional unique constraint for this, but it's you who needs to define it, your entity definition right now just says "here is a user FK and it should be unique", so that is what you get. 

You can use index expression for that - as with postgres you cannot define a conditional/partial unique constraint, but you can define such unique index.

https://stackoverflow.com/questions/16236365/postgresql-conditionally-unique-constraint

So maybe something like this could work (haven't tried it myself):

```ts
// the name is important as that is what will be used for diffing, index expressions are not diffed based on type, but only by name
@Unique({ name: 'myindex', expression: `CREATE UNIQUE INDEX myindex ON identity (user_id) WHERE (type = 'google');` })
```

> this.addSql('alter table "identity" add constraint "identity_type_user_id_unique" unique ("type", "user_id");');

Are you sure about that? I don't think you would get the `type` column in here automatically, I'd say you would have to do `@Unique({ properties: ['type', 'user'] })` to get that.

**Reply by Jcalcaldev** [reply]
This works, small correction though, it has to be an 'Index' decorator.

And you're right on the second point.

Thanks for the suggestion :)
